### DISCUSSION TITLE
Fluent configuration api

### DISCUSSION THREAD

**Discussion by voroninp** [riok/mapperly#1365] [Ideas]
Using `nameof` is fine, but there's a need to specify source and target types every time.
What if there were a static class `Map` which coud be used inside static constructor of the mapper class:

```csharp
public static class Map
{
    [Conditional("xxx")]
    public static void Properties<TSource, TTarget>(Func<TSource, TTarget, (object sourceProperty, object targetProperty)[]> properties)
    {

    }
}

public sealed class Foo
{
    public required int Foo1 { get; init; }
    public required int Foo2 { get; init; }
}

public sealed class Bar
{
    public required int Bar1 { get; init; }
    public required int Bar2 { get; init; }
}

[Mapper]
public sealed partial class Mapper
{
    static Mapper()
    {
        Map.Properties<Foo, Bar>((f, b) => [
            (f.Foo1, b.Bar1), 
            (f.Foo2, b.Bar2)
        ]);
    }

    public partial Bar From(Foo foo);
}
```
Source generater can analyze syntax tree and infer the match between properties.
As method is conditional it will be erased as well.

**Comment by latonz** [comment]
I like the idea of alternative configuration syntaxes, especially ones which allow to get rid of string property names and nameof expressions. However, an approach with method calls to configure the mappings would either need isolation and compilation of the given calls or a complex source analysis. Both would probably result in an explosion of the complexity of Mapperly‘s source code.
Do you have an idea how this could be implemented on Mapperly‘s side?

**Reply by voroninp** [reply]
Yep, the idea was indeed to analyze the syntax tree. I'll play with it a little, and then let you know.

**Reply by latonz** [reply]
I'll close this for now, as we have no concrete plans to support alternative configuration APIs. However, we are still open to new concrete configuration API proposals 😊.
